### PR TITLE
Use N-bit lengths for VZV IndexN format types

### PR DESCRIPTION
**Pull request by Manishearth**
It is impossible for an IndexN array to need more than a length integer of size N, anyway, the max index is always `>=` the length.

Part of #5523

Builds on #5593

We could in theory just have a `VZVFormatCombo<Index, Len>` type that allows free selection, however I'm trying to keep this minimal. Overall the main use case for that is picking things like "a small array of ;argely-sized elements" and we could just expose Index16Len8 for that.  I can see that being useful for things like https://github.com/unicode-org/icu4x/pull/5580, though it also feels like a data microoptimization.

The "total" lines in fingerprints.csv are interspersed in giant diffs, and this basically only gets a max of 2-6 byte wins per data, but the overall data size went down by 200KB. Not amazing, not terrible.

```rust
[18:26:22] मanishearth@manishearth-glaptop2 ~/dev/icu4x/provider/data ^_^ 
$ rg total | awk '{ gsub(/B,/, "", $3); s +=$3} END{print s}' 
23501369
[18:26:08] मanishearth@manishearth-glaptop2 ~/dev/icu4x/provider/data ^_^ 
$ rg total | awk '{ gsub(/B,/, "", $3); s +=$3} END{print s}' 
23391499
```

### CODE REVIEW

**Comment by sffc** [comment]
It's not _super_ common that Index8 is big enough, but it is quite common that Len8 is big enough. I feel like Index8Len16 would be one of the most commonly used settings.

**Comment by Manishearth** [comment]
@sffc we can expose it easily in a followup, yeah. I think Index16Len8 is definitely more useful than Index8 and we should consider looking at where we use Index8 now.

File: utils/zerovec/src/varzerovec/components.rs [hunk 1]
```diff
@@ -28,40 +27,60 @@ pub(super) const MAX_INDEX: usize = u32::MAX as usize;
...
+/// and all of its associated items are hidden from the docs.
 #[allow(clippy::missing_safety_doc)] // no safety section for you, don't implement this trait period
-pub unsafe trait VarZeroVecFormat: 'static + Sized {
+#[doc(hidden)]
+pub unsafe trait IntegerULE: ULE {
     /// The error to show when unable to construct a vec
     #[doc(hidden)]
     const TOO_LARGE_ERROR: &'static str;
+
+    /// Safety: must be sizeof(self)
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/components.rs:54]: Suggestion: Why not just use size_of::<Self>() which is const?

**Review by sffc — APPROVED** [review]

File: utils/zerovec/src/varzerovec/components.rs (hunk 1 above)
Comment (Manishearth) [reply utils/zerovec/src/varzerovec/components.rs:54]: I found it makes the code easier to read but I'm open to that being a followup

Comment (Manishearth) [reply utils/zerovec/src/varzerovec/components.rs:54]: basically `mem::size_of::<F::Index>` feels clunkier than `F::Index::SIZE` in the situations it gets used in.

_however_ in my next PR I'm removing METADATA_WIDTH and that may make the expressions simple enough that a sizeof won't be a big deal

**Pull request merged**